Update build_tree function with SparseKmeans implementation

khoinpd0411 opened the pull request:
## What does this PR do?

Update build_tree function with SparseKmeans implementation and utilize an adaptive clustering method mixing Elkan's and Lloyd's algorithm based on the number of samples.

Improvements:

- Speed up tree construction.
- Resolved convergence issues caused by duplicate samples during clustering
- Introduced an adaptive clustering strategy that dynamically switches between Elkan’s algorithm (for large sample sizes) and Lloyd’s algorithm (for smaller or dense datasets)

## Test CLI & API (`bash tests/autotest.sh`)
Test APIs used by main.py.
- [x] Test Pass
  - (Copy and paste the last outputted line here.)
- [ ] Not Applicable (i.e., the PR does not include API changes.)

## Check API Document
If any new APIs are added, please check if the description of the APIs is added to API document. 
- [ ] API document is updated ([linear](https://www.csie.ntu.edu.tw/~cjlin/libmultilabel/api/linear.html), [nn](https://www.csie.ntu.edu.tw/~cjlin/libmultilabel/api/nn.html))
- [x] Not Applicable (i.e., the PR does not include API changes.)

## Test quickstart & API (`bash tests/docs/test_changed_document.sh`)
If any APIs in quickstarts or tutorials are modified, please run this test to check if the current examples can run correctly after the modified APIs are released.

khoinpd0411 committed:
Update build_tree function with sparsekmeans and utilize an adaptive clustering method mixing Elkan's and Lloyd's algorithm based on the number of samples

Eleven1Liu commented on requirements.txt:
```diff
@@ -6,3 +6,4 @@ scikit-learn
 scipy<1.14.0
 tqdm
 psutil
+sparsekmeans
```
Please also add sparsekmeans to `install_requires` https://github.com/ntumlgroup/LibMultiLabel/blob/a0bef91d770d449dce6cc5a1a0fb8dc1c9d0647e/setup.cfg#L27-L35

Eleven1Liu replied:
~~Please bump version to 0.8.0~~
https://github.com/ntumlgroup/LibMultiLabel/blob/a0bef91d770d449dce6cc5a1a0fb8dc1c9d0647e/setup.cfg#L3

zhi-bao replied:
Sparsekmeans requires **Python >= 3.10**, whereas LibMultiLabel supports **Python >= 3.8**.
This causes installation issues when users try to install LibMultiLabel in Python 3.8 or 3.9.
There are two approach for this issue:
- Update LibMultiLabel to require Python >= 3.10. 
- Or, detect the user's environment and apply the corresponding workaround. 
There is much room for discussion.

Eleven1Liu replied:
@khoinpd0411 No need to bump version now, we will release with #20.

chcwww commented on libmultilabel/linear/tree.py:
```diff
@@ -277,24 +277,37 @@ def _build_tree(label_representation: sparse.csr_matrix, label_map: np.ndarray,
+                tol=0.0001,
+                random_state=np.random.randint(2**31 - 1),
+                verbose=True
+                )
```
I'm just wondering why the indentation isn't aligned with `line:287`.
BTW, should we pass the `verbose` flag through `_build_tree()` so that we can control the output when training?
And would it be better to do something like (I'm not sure)
```python
if label_representation.shape[0] > 10000:
    kmeans_algo = ElkanKmeans
else:
    kmeans_algo = LloydKmeans
kmeans = kmeans_algo(those_params)
```

Eleven1Liu replied:
For the formatting issue, please use [black formatter](https://pypi.org/project/black/).

maclin726 commented on libmultilabel/linear/tree.py:
```diff
@@ -277,24 +277,37 @@ def _build_tree(label_representation: sparse.csr_matrix, label_map: np.ndarray,
+            child = _build_tree(child_representation, child_map, d + 1, K, dmax)
+        else:
+            child = Node(label_map=child_map, children=[])
+        
```
Maybe it's better to have

```python
num_unique_labels = len(np.unique(metalabels))
if len(num_unique_labels) == K:
    children = []
    for i in range(K):
        child_representation = label_representation[metalabels == i]
        child_map = label_map[metalabels == i]
        child = _build_tree(child_representation, child_map, d + 1, K, dmax)
        children.append(child)
else:
    children = [
        Node(label_map=label_map[metalabels == i], children=[])
            for i in range(num_unique_labels)
    ]
```

khoinpd0411 committed:
Refactor _build_tree function:
- Eliminate duplicated declarations of KMeans parameters for Lloyd’s and Elkan’s methods.
- Move the check for the number of unique labels outside the loop.
- Combine terminal conditions where d >= dmax and the number of unique samples is less than K.

khoinpd0411 committed:
Merge remote-tracking branch 'upstream/master' into build-tree-sparsekmeans

khoinpd0411 committed:
Add sparsekmeans as a dependency in setup.cfg

Eleven1Liu reviewed (approved):
verified by @maclin726
